### PR TITLE
Miscellaneous

### PR DESCRIPTION
**Pull request by mahasajid**
Bonus features

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -10,8 +10,8 @@ gem "rails", "~> 7.0.3"
 gem "sprockets-rails"
 
 # Use sqlite3 as the database for Active Record
-gem "sqlite3", "~> 1.4"
-
+#gem "sqlite3", "~> 1.4"
+gem "pg"
```
Comment (salman-haider-confiz) [inline Gemfile:14]: Always Remove Commented code before commit

File: Gemfile [hunk 2]
```diff
@@ -71,3 +71,23 @@ group :test do
...
+gem 'hiredis'
+gem 'redis'
+gem 'dotenv-rails'
+gem "omniauth"
+gem "omniauth-google-oauth2"
+gem "omniauth-rails_csrf_protection"
+gem 'figaro'
+gem 'rake'
+gem 'bonsai-elasticsearch-rails'
+gem 'faker'
```
Comment (salman-haider-confiz) [inline Gemfile:93]: Always lock your Gems by adding the Gem version.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,15 +1,18 @@
 class ApplicationController < ActionController::Base
 
 include CurrentCart
-before_action :set_cart
+before_action :set_cart , :initialize_products
 #before_action :set_line_item
```
Comment (salman-haider-confiz) [inline app/controllers/application_controller.rb]: Remove commented code.

File: app/controllers/carts_controller.rb [hunk 4]
```diff
@@ -1,5 +1,8 @@
 class CartsController < ApplicationController
-  before_action :set_cart, only: %i[ show edit update destroy  confirm ]
+
+#include CurrentCart
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb:3]: Same as above.

File: app/controllers/carts_controller.rb [hunk 5]
```diff
@@ -1,5 +1,8 @@
...
+before_action :set_cart, only: %i[ show edit update destroy  confirm ]
+before_action :update_order_status, only: %i[update]
   #before_action :set_line_item, only: %i[ checkout ]
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb]: Same as above

File: app/controllers/carts_controller.rb [hunk 6]
```diff
@@ -11,20 +14,29 @@ def show
   end
 
   # GET /carts/new
-  def new
-    @cart = Cart.new
+   def new
+  #@cart = Cart.new
+  #@cart.user = current_user
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb]: This should be removed

File: app/controllers/carts_controller.rb [hunk 7]
```diff
@@ -11,20 +14,29 @@ def show
   end
 
   # GET /carts/new
-  def new
-    @cart = Cart.new
+   def new
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb:16]: I think you should add that user should be logged in before creating cart.

File: app/controllers/carts_controller.rb [hunk 8]
```diff
@@ -11,20 +14,29 @@ def show
...
 
   # POST /carts or /carts.json
   def create
-    @cart = Cart.new(cart_params)
-    @cart.save
+
+    @cart = current_user.cart.build(cart_params)
+    @cart.user = current_user
+    # @cart = Cart.new(cart_params)
+    # @cart.save
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb]: Remove comments.

File: app/controllers/carts_controller.rb [hunk 9]
```diff
@@ -39,17 +51,29 @@ def create
...
-   #@friend = current_account.friends.build(friend_params)
+   
    #@user = current_user
 
-   @cart.order_status = "ordered"
+   #params[:order_status] = "ordered"
+   #@cart.order_status = "ordered"
    #current_user.cart.build(@cart_params)
+   #changed_order_status("ordered")
+
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb]: Remove commented Code.

File: app/controllers/carts_controller.rb [hunk 10]
```diff
@@ -69,15 +93,17 @@ def destroy
...
   end
 
 
 #PUT /checkout
 def confirm
+
   respond_to do |format|
     if @cart.update(cart_params)
-      format.html { redirect_to products_path, notice: "Order was placed" }
+
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb:83]: Remove Extra Lines as well.

File: app/controllers/carts_controller.rb [hunk 11]
```diff
@@ -98,15 +124,22 @@ def set_cart
     def cart_params
       #params.fetch(:cart, {})
 
-      params.require(:cart).permit( :user_id, :order_status, :users_attributes => [:id, :shipping_address, :name, :email])
+      params.require(:cart).permit( :user_id, :order_status,:shipping_address, :billing_address)
     end
 
     # def user_params
     #   params.require(:user).permit( :name , :shipping_address, :email )
     # end
```
Comment (salman-haider-confiz) [inline app/controllers/carts_controller.rb]: Remove it.

File: app/controllers/search_controller.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
+class SearchController < ApplicationController
+    
+    def search
+    #   unless params[:query].blank?
+    #     @products = Product.search( params[:query] )
+    #   end
+
```
Comment (salman-haider-confiz) [inline app/controllers/search_controller.rb]: Remove it.

File: app/models/cart.rb [hunk 13]
```diff
@@ -23,4 +25,38 @@ def total_price
...
+#     item.product.save
+#     }
+
+#   else
+
+#     self.line_items.each {|item|
+#     item.product.stock = item.product.stock + item.quantity
+#     item.product.save
+#     }
+
```
Comment (salman-haider-confiz) [inline app/models/cart.rb]: Remove it.

File: app/models/concerns/current_cart.rb [hunk 14]
```diff
@@ -5,9 +5,39 @@ module CurrentCart
...
+    #         session[:cart_id] = nil
+    #       end
+    #     end
+  
+    #     if cart.order_status == "ordered" || session[:cart_id] == nil
+            
+    #       @current_cart = Cart.create
+    #       session[:cart_id] = @current_cart.id
+    #     end
+    #   end
```
Comment (salman-haider-confiz) [inline app/models/concerns/current_cart.rb]: Remove it

**Review by salman-haider-confiz — CHANGES_REQUESTED** [review]

File: app/controllers/carts_controller.rb (hunk 7 above)
Comment (mahasajid) [reply app/controllers/carts_controller.rb:16]: authenticate_user! used in application controller

**Pull request merged**